### PR TITLE
Update Fibonacci.java

### PR DESCRIPTION
**Pull request by XudongX**
Using HashMap will cause ConcurrentModificationException, we should use ConcurrentHashMap instead.

### CODE REVIEW

**Comment by xudong07** [comment]
I am a beginner in java. I don't know why, but it actually appeared.

This is my environment:
```
java 11 2018-09-25
Java(TM) SE Runtime Environment 18.9 (build 11+28)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11+28, mixed mode)
```


This is my main method:
```
    public static void main(String[] args) {
        Fibonacci fibonacci = new Fibonacci();
        fibonacci.fibonacci(5);
    }
```

This is stack trace infomation:
```
Exception in thread "main" java.util.ConcurrentModificationException
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1134)
	at com.insightfullogic.java8.answers.chapter5.Fibonacci.fibonacci(Fibonacci.java:17)
	at com.insightfullogic.java8.answers.chapter5.Fibonacci.lambda$fibonacci$0(Fibonacci.java:17)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at com.insightfullogic.java8.answers.chapter5.Fibonacci.fibonacci(Fibonacci.java:17)
	at com.insightfullogic.java8.answers.chapter5.Fibonacci.lambda$fibonacci$0(Fibonacci.java:17)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at com.insightfullogic.java8.answers.chapter5.Fibonacci.fibonacci(Fibonacci.java:17)
	at com.insightfullogic.java8.answers.chapter5.Fibonacci.main(Fibonacci.java:22)
```

Should I try to use jdk1.8?

**Comment by RichardWarburton** [comment]
Oh interesting. Yeah I was testing on Java 8. It might be that the behaviour here has changed in 11.

**Comment by xudong07** [comment]
It must be a complex problem 😭, and I find a relative discussion in [stackoverflow](https://stackoverflow.com/questions/28840047/recursive-concurrenthashmap-computeifabsent-call-never-terminates-bug-or-fea). I can't figure it out, but it may be helpful.

File: src/main/java/com/insightfullogic/java8/answers/chapter5/Fibonacci.java [hunk 1]
```diff
@@ -8,7 +8,7 @@
     private final Map<Integer,Long> cache;
 
     public Fibonacci() {
-        cache = new HashMap<>();
+        cache = new ConcurrentHashMap<>();
```
Comment (takingsss12) [inline src/main/java/com/insightfullogic/java8/answers/chapter5/Fibonacci.java]: `ConcurrentHashMap` is useless in this context as the fibonacci sequence is evaluated inside a single thread.

Remember, recursive evaluation has no relationship with parallel computing.

Comment (tweimer) [reply src/main/java/com/insightfullogic/java8/answers/chapter5/Fibonacci.java:11]: I does help, I just checked it with Java 14. HashMap doesn't work while ConcurrentHashMap does. See my comment below.

**Comment by tweimer** [comment]
I checked this, and I verified that the original code works in Java 8 while it fails in Java 14 in my case.

Actually, the documentation of the computeIfAbsent() in HashMap states:

`@throws ConcurrentModificationException if it is detected that the mapping function modified this map`
That's evidently the case here.

The ConcurrentHashMap even states: "The mapping function **must not modify** this map during computation.", but it actually works. Still, this solution is a bit questionable IMHO.

@RichardWarburton Do you merge this PR?

**Review by tweimer — CHANGES_REQUESTED** [review]
This doesn't even compile.

**Comment by tweimer** [comment]
On another try, this doesn't really fix it. it works with `fibonacci(11)` but fails with `fibonacci(12)`. 

```
Exception in thread "main" java.lang.IllegalStateException: Recursive update
	at java.base/java.util.concurrent.ConcurrentHashMap.transfer(ConcurrentHashMap.java:2552)
	at java.base/java.util.concurrent.ConcurrentHashMap.addCount(ConcurrentHashMap.java:2354)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1776)
```

**Comment by RichardWarburton** [comment]
Indeed. I think this is an exercise that initially made sense with Java 8 but has aged poorly with time. I don't think that the switch to ConcurrentHashMap is a good alternative. It might just be worth removing the exercise altogether.

**Comment by tweimer** [comment]
I agree with removing it :+1:

**Comment by XudongX** [comment]
Thank you guys, I’ve realized that `concurrentHashMap` does no help in the code. Maybe I need to close this pull request.

**Pull request closed without merging**